### PR TITLE
Add template functions for list/inspect message

### LINKED ISSUE #614: Add command to show the start message again
### Description

In case you missed it and it scrolled off, or if you are doing it from a GUI and want to show a dialog to the user.

Currently the information is shown with the rest of the clutter, which can make it hard to screen scrape (or read)

```
INFO[0000] Using the existing instance "docker"         
INFO[0000] [hostagent] Starting QEMU (hint: to watch the boot progress, see "/home/anders/.lima/docker/serial.log") 
INFO[0000] SSH Local Port: 43151                        
INFO[0000] [hostagent] Waiting for the essential requirement 1 of 5: "ssh" 
INFO[0021] [hostagent] The essential requirement 1 of 5 is satisfied 
INFO[0021] [hostagent] Waiting for the essential requirement 2 of 5: "user session is ready for ssh" 
INFO[0021] [hostagent] The essential requirement 2 of 5 is satisfied 
INFO[0021] [hostagent] Waiting for the essential requirement 3 of 5: "sshfs binary to be installed" 
INFO[0021] [hostagent] The essential requirement 3 of 5 is satisfied 
INFO[0022] [hostagent] Waiting for the essential requirement 4 of 5: "/etc/fuse.conf to contain \"user_allow_other\"" 
INFO[0022] [hostagent] The essential requirement 4 of 5 is satisfied 
INFO[0022] [hostagent] Waiting for the essential requirement 5 of 5: "the guest agent to be running" 
INFO[0022] [hostagent] The essential requirement 5 of 5 is satisfied 
INFO[0022] [hostagent] Mounting "/home/anders"          
INFO[0022] [hostagent] Mounting "/tmp/lima"             
INFO[0022] [hostagent] Waiting for the optional requirement 1 of 1: "user probe 1/1" 
INFO[0022] [hostagent] Forwarding "/run/user/1000/docker.sock" (guest) to "/home/anders/.lima/docker/sock/docker.sock" (host) 
INFO[0022] [hostagent] Forwarding "/run/lima-guestagent.sock" (guest) to "/home/anders/.lima/docker/ga.sock" (host) 
INFO[0022] [hostagent] Not forwarding TCP 127.0.0.53:53 
INFO[0022] [hostagent] The optional requirement 1 of 1 is satisfied 
INFO[0022] [hostagent] Not forwarding TCP 0.0.0.0:22    
INFO[0022] [hostagent] Not forwarding TCP [::]:22       
INFO[0022] [hostagent] Waiting for the final requirement 1 of 1: "boot scripts must have finished" 
INFO[0025] [hostagent] The final requirement 1 of 1 is satisfied 
INFO[0025] READY. Run `limactl shell docker` to open the shell. 
INFO[0025] Message from the instance "docker":          
To run `docker` on the host (assumes docker-cli is installed):
$ export DOCKER_HOST=unix:///home/anders/.lima/docker/sock/docker.sock
$ docker ...
```

Something like `limactl message` to show the message again, perhaps ? I could just filter out the INFO, otherwise.

### PR DESCRIPTION
**Pull request by afbjorklund**
Makes it possible to show messages using a format string like:

    {{.Name}}{{printf "\n"}}{{.Message | indent 2 | trim "\n"}}

Without having to add a command like `message` or a flag like `--message`.

Sample output:

```console
$ limactl ls -f '{{.Name}}{{printf "\n"}}{{.Message | indent 2 | trim "\n"}}'
k3d
  To run `kubectl` on the host (assumes kubectl is installed), run the following commands:
  ------
  mkdir -p "/home/anders/.lima/k3d/conf"
  export KUBECONFIG="/home/anders/.lima/k3d/conf/kubeconfig.yaml"
  limactl shell k3d sh -c 'cat ~/.kube/config' >$KUBECONFIG
  kubectl ...
  ------
  To run `docker` on the host (assumes docker-cli is installed), run the following commands:
  ------
  export DOCKER_HOST="unix:///home/anders/.lima/k3d/sock/docker.sock"
  docker ...
  ------
kind
  To run `kubectl` on the host (assumes kubectl is installed), run the following commands:
  ------
  mkdir -p "/home/anders/.lima/kind/conf"
  export KUBECONFIG="/home/anders/.lima/kind/conf/kubeconfig.yaml"
  limactl shell kind sh -c 'cat ~/.kube/config' >$KUBECONFIG
  kubectl ...
  ------
  To run `docker` on the host (assumes docker-cli is installed), run the following commands:
  ------
  export DOCKER_HOST="unix:///home/anders/.lima/kind/sock/docker.sock"
  docker ...
  ------
```

Closes #614

### CODE REVIEW

**Comment by afbjorklund** [comment]
And another template function, to be able to decorate instances without a message (as show in example)

`limactl ls -f '{{.Name}}{{printf "\n"}}{{.Message | missing "<no message>" | indent 2 | trim "\n"}}'`

File: cmd/limactl/list.go [hunk 1]
```diff
@@ -126,7 +149,12 @@ func listAction(cmd *cobra.Command, args []string) error {
 	}
 
 	if goFormat != "" {
-		tmpl, err := template.New("format").Parse(goFormat)
+		funcMap := template.FuncMap{
+			"indent":  indentString,
+			"trim":    trimString,
+			"missing": missingString,
+		}
```
Comment (AkihiroSuda) [inline cmd/limactl/list.go:156]: Can we move this to under `pkg/`, and use this for `limactl inspect` too?

Comment (AkihiroSuda) [reply cmd/limactl/list.go:156]: I'd also like to see unit tests

Comment (jandubois) [reply cmd/limactl/list.go:156]: Also documentation; not sure where to put it though

**Comment by jandubois** [comment]
I wonder if it is possible to provide default values for the 2nd argument, so you could write

```
limactl ls -f '{{.Name}}{{"\n"}}{{.Message | missing | indent | trim }}'
```

instead of[^1]

```
limactl ls -f '{{.Name}}{{printf "\n"}}{{.Message | missing "<missing>" | indent 2 | trim "\n"}}'
```

Maybe using `i ...interface{}` for the second argument can make it work?

Of course we then need a round of bikeshedding to agree on the right default values. E.g. I think `trim` should trim all whitespace and not just newlines.

[^1]: You can use `{{"\n"}}` instead of `{{printf "\n"}}`.

**Comment by afbjorklund** [comment]
The simple approach does **not** work, at least...

`template: format:1:10: executing "format" at <indent>: wrong number of args for indent: want 2 got 1`

> Of course we then need a round of bikeshedding to agree on the right default values.

Like there is no argument about indent sizes... ;-)

>  think trim should trim all whitespace and not just newlines.

Trimming on all whitespace, removes the indent ?

File: cmd/limactl/list.go (hunk 1 above)
Comment (afbjorklund) [reply cmd/limactl/list.go:156]: > Also documentation; not sure where to put it though

Currently the documentation is duplicated in the cmds

`--format '{{ <go template> }}' - if the format begins and ends with '{{ }}', then it is used as a go template.`

Comment (afbjorklund) [reply cmd/limactl/list.go:156]: Maybe something like `--list-fields`, like `--list-funcs` ?

Which by the way seems to be missing from `inspect'

**Comment by jandubois** [comment]
> The simple approach does **not** work, at least...
> 
> `template: format:1:10: executing "format" at <indent>: wrong number of args for indent: want 2 got 1`

It worked for me with:

```golang
// IndentString add spaces to beginning of each line
func IndentString(i ...interface{}) string {
	size := 2
	if len(i) > 1 {
		size = i[0].(int)
	}
	text := i[len(i)-1].(string)
	prefix := strings.Repeat(" ", size)
	return PrefixString(prefix, text)
}
```

Of course needs proper type checking to avoid it from panicking.

```console
$ limactl ls -f '{{.Name}}{{"\n"}}{{.Dir | indent | trim "\n"}}'
alpine
  /Users/jan/.lima/alpine
default
  /Users/jan/.lima/default

$ limactl ls -f '{{.Name}}{{"\n"}}{{.Dir | indent 4 | trim "\n"}}'
alpine
    /Users/jan/.lima/alpine
default
    /Users/jan/.lima/default
```

**Comment by jandubois** [comment]
The need for the `trim` function seems to come from the fact that `PrefixString` will append a trailing newline, even when the input string didn't have one. I think with this change, we can get rid of the `trim` function:

```diff
--- pkg/templateutil/templateutil.go
+++ pkg/templateutil/templateutil.go
@@ -8,17 +8,32 @@ import (
 // PrefixString adds prefix to beginning of each line
 func PrefixString(prefix string, text string) string {
        result := ""
+       trailingNewline := strings.HasSuffix(text, "\n")
        for _, line := range strings.Split(text, "\n") {
                if line == "" {
                        continue
                }
                result += prefix + line + "\n"
        }
-       return result
+       if trailingNewline {
+               return result
+       }
+       return strings.TrimSuffix(result, "\n")
 }
```

**Comment by jandubois** [comment]
> > Of course we then need a round of bikeshedding to agree on the right default values.
> Like there is no argument about indent sizes... ;-)

Suggestions for defaults:

* `indent`: `2`

* `missing`: `<missing>`

* `prefix`: `> `

An alternative for `missing` would be `<no value>` which the template package uses to render `nil`.

For `prefix` my first though would be the empty string, but that makes it a no-op. Or 2 spaces, but that makes it an alias for `indent`. So `> ` is kind of a fallback. I guess I don't really see the need for `prefix`...

**Comment by jandubois** [comment]
> Of course needs proper type checking to avoid it from panicking.

Doesn't actually panic:

```console
$ limactl ls -f '{{.Name}}{{"\n"}}{{2 | indent 2}}'
alpine
FATA[0000] template: format:1:23: executing "format" at <indent 2>: error calling indent: interface conversion: interface {} is int, not string

$ limactl ls -f '{{.Name}}{{"\n"}}{{.Dir | indent "x"}}'
alpine
FATA[0000] template: format:1:26: executing "format" at <indent "x">: error calling indent: interface conversion: interface {} is string, not int
```

So maybe the code is good enough without additional type checking.

**Comment by jandubois** [comment]
But maybe it is more user-friendly to provide custom errors:

```golang
// IndentString add spaces to beginning of each line
func IndentString(i ...interface{}) (string, error) {
	if len(i) == 0 {
		return "", errors.New("function takes at at least one string argument")
	}
	if len(i) > 2 {
		return "", errors.New("function takes at most 2 arguments")
	}
	var ok bool
	size := 2
	if len(i) > 1 {
		if size, ok = i[0].(int); !ok {
			return "", errors.New("optional first argument must be an integer")
		}
	}
	text := ""
	if text, ok = i[len(i)-1].(string); !ok {
		return "", errors.New("last argument must be a string")
	}
	prefix := strings.Repeat(" ", size)
	return PrefixString(prefix, text), nil
}
```

Resulting in:

```console
$ limactl ls -f '{{indent}}'
FATA[0000] template: format:1:2: executing "format" at <indent>: error calling indent: function takes at at least one string argument

$ limactl ls -f '{{2 | indent}}'
FATA[0000] template: format:1:6: executing "format" at <indent>: error calling indent: last argument must be a string

$ limactl ls -f '{{.Dir | indent "x"}}'
FATA[0000] template: format:1:9: executing "format" at <indent "x">: error calling indent: optional first argument must be an integer

$ limactl ls -f '{{.Dir | indent 4 8}}'
FATA[0000] template: format:1:9: executing "format" at <indent 4 8>: error calling indent: function takes at most 2 arguments
```

**Comment by afbjorklund** [comment]
> The need for the `trim` function seems to come from the fact that `PrefixString` will append a trailing newline, even when the input string didn't have one.

The message had a trailing newline, and then the --format added another newline. So that's why it was trimmed.

Otherwise there was an empty space, between instances.

```
foo
  ...

bar
  ...

baz
  <no message>
```

**Comment by jandubois** [comment]
> The message had a trailing newline, and then the --format added another newline. So that's why it was trimmed.

I see. I don't think the trailing newline on the `message` is ever useful when evaluated in the `--format` context, so I would prefer to always trim it before evaluating the template:

```diff
@@ -289,6 +286,7 @@ func PrintInstances(w io.Writer, instances []*Instance, format string) error {
                        return fmt.Errorf("invalid go template: %w", err)
                }
                for _, instance := range instances {
+                       instance.Message = strings.TrimSuffix(instance.Message, "\n")
                        err = tmpl.Execute(w, instance)
                        if err != nil {
                                return err
```

My goal is to have the templates do what the user want automatically, while keeping the templates as concise as possible.

```console
$ limactl ls -f '{{.Name}}{{printf "\n"}}{{.Message | missing | indent }}'
alpine
  <missing>
default
  <missing>
docker
  To run `docker` on the host (assumes docker-cli is installed), run the following commands:
  ------
  docker context create lima-docker --docker "host=unix:///Users/jan/.lima/docker/sock/docker.sock"
  docker context use lima-docker
  docker run hello-world
  ------
podman
  To run `podman` on the host (assumes podman-remote is installed), run the following commands:
  ------
  podman system connection add lima-podman "unix:///Users/jan/.lima/podman/sock/podman.sock"
  podman system connection default lima-podman
  podman run quay.io/podman/hello
  ------
```

Can you think of a use-case where you would want to keep the trailing newline?

**Comment by afbjorklund** [comment]
Apparently it is a YAML syntax thing, if we wanted the message to **not** have a newline it should use `|-`.

```
# Message. Information to be shown to the user, given as a Go template for the instance.
# The same template variables as for listing instances can be used, for example {{.Dir}}.
# You can view the complete list of variables using `limactl list --list-fields` command.
# It also includes {{.HostOS}} and {{.HostArch}} vars, for the runtime GOOS and GOARCH.
# 🟢 Builtin default: null
# message: |
#   This will be shown to the user.
```

Now we use the template also to show json and yaml (since 820b78761946dd4b3902536e98ec949c876e0fc3) so otherwise it would be lying ?

**Comment by afbjorklund** [comment]
> For `prefix` my first though would be the empty string, but that makes it a no-op. Or 2 spaces, but that makes it an alias for `indent`. So `> ` is kind of a fallback. I guess I don't really see the need for `prefix`...

Removed the func map, it was more of an implementation detail anyway.

**Comment by afbjorklund** [comment]
Did the rebase, but probably want to merge "textutil" (no tests?) and "templateutil"

**Comment by jandubois** [comment]
> Now we use the template also to show json and yaml (since [820b787](https://github.com/lima-vm/lima/commit/820b78761946dd4b3902536e98ec949c876e0fc3)) so otherwise it would be lying ?

I would not consider it "lying"; it is just that we can define `.Message` to be the expanded template, with any trailing newlines stripped. I believe there never is any reason to keep the trailing newlines, but I may be wrong. Please show any example where they would be useful.

Note the you still have `.Config.Message` containing the original template. I am **not** suggesting that we change this value, only the already processed `.Message` one.

File: pkg/templateutil/templateutil.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package templateutil
```
Comment (AkihiroSuda) [inline pkg/templateutil/templateutil.go:1]: This should be merged to `pkg/textutil`.

Comment (AkihiroSuda) [reply pkg/templateutil/templateutil.go:1]: ("templateutil" is reserved for `template://TEMPLATE` utils)

**Comment by jandubois** [comment]
Conversation here seems to have stalled. From my point of view, I would like to see the `indent` and `missing` function to support default arguments (`2`, and `"missing"`, respectively).

I also think that `PrefixString` should not append an extra newline if the input value didn't have one, and that the `.Message` field should automatically be stripped of trailing newlines. I think this would completely eliminate the need for the `trim` function, which I would like to remove unless there is some demonstrated use-case.

With those changes I would be willing to approve this PR.

Let me know if you need help, and I can propose a commit that implements the changes I would like to see.

Or continue the conversation, if you think my suggestions are problematic.

**Comment by afbjorklund** [comment]
I put it on the backlog, since the feature wasn't that important...

* #614

But will move to the new "util" directory, and add the new arguments.

We can strip the newline in Execute, but it did feel like a bit of a hack.

**Comment by afbjorklund** [comment]
Added the default arguments, with the custom argument checker and errors.

The existing functions were a bit weird, since they call `panic` instead of errors.

Also seemed like they had hardcoded some stuff for list, so less generic ?

e.g. the JSON adds a trailing newline and  the YAML adds a document marker...

**Review by jandubois — APPROVED** [review]
I'm fine with merging this PR as-is, and addressing any further feedback in a follow-up PR.

### Feedback:

I expected the default `missing` argument to be `<missing>`, as `<no message>` only makes sense when the final argument is `.Message`. However, I expect this to be the primary use case of `missing`, so maybe this is fine.

It just occurred to me that we could simply set `.Message` to `<no message>` if it is empty, eliminating the need to use the `missing` function in this context. Not sure if we can agree on this though.

I also just noticed that the `limactl ls --help` output is badly formatted (separate issue, not affected by this PR):

```console
$ limactl ls --help
List instances of Lima.
		The output can be presented in one of several formats, using the --format <format> flag.

		  --format json  - output in json format
		  --format yaml  - output in yaml format
		  --format table - output in table format
		  --format '{{ <go template> }}' - if the format begins and ends with '{{ }}', then it is used as a go template.

		These functions are available to go templates:

		  indent <size>: add spaces to beginning of each line
		  missing <message>: return message if the text is empty

		The following legacy flags continue to function:
		  --json - equal to '--format json'

Usage:
  limactl list [flags] [INSTANCE]...

Aliases:
  list, ls

Flags:
  -f, --format string   output format, one of: json, yaml, table, go-template (default "table")
  -h, --help            help for list
      --json            JSONify output
      --list-fields     List fields available for format
  -q, --quiet           Only show names

Global Flags:
      --debug   debug mode
```

The first block of text (before the actual "Usage:") is indented by 16 spaces (2 tabs), whereas it should be indented just 2 spaces.

**Comment by afbjorklund** [comment]
Changed the default missing text (to `missing`), I just wanted to show a single message anyway.

Should remove the two tabs then, so that the generated func help lines up with the static text.

**Comment by afbjorklund** [comment]
We can do it separately, just that this PR adds two tabs - that would also need to be changed, accordingly

**Comment by jandubois** [comment]
> We can do it separately, just that this PR adds two tabs - that would also need to be changed, accordingly

Ok, let's do it separately, so that we can draw down the big backlog of in-flight PRs.

**Pull request merged**